File: examples/typescript/part-7/22-supply-chain-security.ts
// Illustrative examples for notes/part-7-testing-compliance/22-supply-chain-security.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).
/// <reference path="../types.d.ts" />

import { createHash } from "node:crypto";

enum ArtifactType {
  Dependency = "dependency",
  Container = "container",
  Prompt = "prompt",
  Policy = "policy",
  Tool = "tool",
  MCPServer = "mcp_server",
  Model = "model",
  Dataset = "dataset",
}

interface Artifact {
  name: string;
  type: ArtifactType;
  version: string;
  hash: string;
  source: string;
  owner: string;
  reviewed: boolean;
  capabilities?: string[];
}

function validateInventory(items: Artifact[]): void {
  for (const item of items) {
    if (!item.name || !item.type || !item.version) {
      throw new Error("artifact has required empty fields");
    }

    if (!item.owner) {
      throw new Error(`artifact has no owner: ${item.name}`);
    }

    if (!item.hash && item.type !== ArtifactType.Model) {
      throw new Error(`artifact has no hash: ${item.name}`);
    }

    if (isCapabilityArtifact(item.type) && !item.reviewed) {
      throw new Error(`capability artifact is not reviewed: ${item.name}`);
    }
  }
}

function isCapabilityArtifact(artifactType: ArtifactType): boolean {
  return (
    artifactType === ArtifactType.Tool ||
    artifactType === ArtifactType.MCPServer ||
    artifactType === ArtifactType.Policy ||
    artifactType === ArtifactType.Prompt
  );
}

function hashBytes(data: string | Uint8Array): string {
  const input =
    typeof data === "string" ? data : new TextDecoder().decode(data);
  return createHash("sha256").update(input).digest("hex");
}

function newPromptArtifact(
  name: string,
  version: string,
  source: string,
  owner: string,
  content: Uint8Array,
): Artifact {
  return {
    name,
    type: ArtifactType.Prompt,
    version,
    hash: hashBytes(content),
    source,
    owner,
    reviewed: true,
  };
}

interface ToolSchema {
  name: string;
  version: string;
  schema: Record<string, unknown>;
  hash: string;
}

function computeSchemaHash(schema: Record<string, unknown>): string {
  return hashBytes(JSON.stringify(schema));
}

function validateToolSchema(toolSchema: ToolSchema): void {
  const actual = computeSchemaHash(toolSchema.schema);
  if (actual !== toolSchema.hash) {
    throw new Error(`tool schema hash mismatch: ${toolSchema.name}`);
  }
}

class Allowlist {
  allowed: Record<string, string>; // name -> hash

  constructor(allowed: Record<string, string>) {
    this.allowed = allowed;
  }

  check(item: Artifact): void {
    const expectedHash = this.allowed[item.name];
    if (!expectedHash) {
      throw new Error(`artifact not allowlisted: ${item.name}`);
    }

    if (expectedHash !== item.hash) {
      throw new Error(`artifact hash mismatch: ${item.name}`);
    }
  }
}

File: examples/python/part-7/22-supply-chain-security.py
# Illustrative examples for notes/part-7-testing-compliance/22-supply-chain-security.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List


class ArtifactType(str, Enum):
    DEPENDENCY = "dependency"
    CONTAINER = "container"
    PROMPT = "prompt"
    POLICY = "policy"
    TOOL = "tool"
    MCP_SERVER = "mcp_server"
    MODEL = "model"
    DATASET = "dataset"


@dataclass
class Artifact:
    name: str
    type: ArtifactType
    version: str
    hash: str = ""
    source: str = ""
    owner: str = ""
    reviewed: bool = False
    capabilities: List[str] = field(default_factory=list)


def validate_inventory(items: List[Artifact]) -> None:
    for item in items:
        if not item.name or not item.type or not item.version:
            raise ValueError("artifact has required empty fields")

        if not item.owner:
            raise ValueError(f"artifact has no owner: {item.name}")

        if not item.hash and item.type != ArtifactType.MODEL:
            raise ValueError(f"artifact has no hash: {item.name}")

        if _is_capability_artifact(item.type) and not item.reviewed:
            raise ValueError(f"capability artifact is not reviewed: {item.name}")


def _is_capability_artifact(artifact_type: ArtifactType) -> bool:
    return artifact_type in (
        ArtifactType.TOOL,
        ArtifactType.MCP_SERVER,
        ArtifactType.POLICY,
        ArtifactType.PROMPT,
    )


def hash_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def new_prompt_artifact(
    name: str, version: str, source: str, owner: str, content: bytes
) -> Artifact:
    return Artifact(
        name=name,
        type=ArtifactType.PROMPT,
        version=version,
        hash=hash_bytes(content),
        source=source,
        owner=owner,
        reviewed=True,
    )


@dataclass
class ToolSchema:
    name: str
    version: str
    schema: Dict[str, Any]
    hash: str


def compute_schema_hash(schema: Dict[str, Any]) -> str:
    encoded = json.dumps(schema, sort_keys=True).encode("utf-8")
    return hash_bytes(encoded)


def validate_tool_schema(tool_schema: ToolSchema) -> None:
    actual = compute_schema_hash(tool_schema.schema)
    if actual != tool_schema.hash:
        raise ValueError(f"tool schema hash mismatch: {tool_schema.name}")


@dataclass
class Allowlist:
    allowed: Dict[str, str]  # name -> hash

    def check(self, item: Artifact) -> None:
        expected_hash = self.allowed.get(item.name)
        if expected_hash is None:
            raise ValueError(f"artifact not allowlisted: {item.name}")

        if expected_hash != item.hash:
            raise ValueError(f"artifact hash mismatch: {item.name}")
